modyn/supervisor/internal/pipeline_executor/evaluation_executor.py

```diff
@@ -0,0 +1,331 @@
+            The logs of the evaluations.
+        """
+        assert self.grpc.evaluator is not None, "Evaluator not initialized."
+        assert self.pipeline.evaluation is not None, "Evaluation config not set."
+        pipeline_status_queue.put(
+            pipeline_stage_msg(PipelineStage.EVALUATE, MsgType.ID, id_submsg(IdType.TRIGGER, trigger_id))
+        )
+
+        eval_requests: list[EvalRequest] = []
+
+        for eval_handler in self.eval_handlers:
+            if eval_handler.config.execution_time != "after_training":
+                continue
+
+            handler_eval_requests = eval_handler.get_eval_requests_after_training(
+                trigger_id=trigger_id,
+                training_id=training_id,
+                model_id=model_id,
+                training_interval=(first_timestamp, last_timestamp),
+            )
+            eval_requests += handler_eval_requests
+
+        num_workers = None
+        if self.config.supervisor:
+            num_workers = self.config.supervisor.after_training_evaluation_workers
+        logs = self._launch_evaluations_async(eval_requests, log, eval_status_queue, num_workers)
+        return logs
+
+    def run_post_pipeline_evaluations(self, eval_status_queue: Queue) -> SupervisorLogs:
+        """Evaluate the trained models after the core pipeline and store the results."""
+        if not self.pipeline.evaluation:
+            return SupervisorLogs(stage_runs=[])
+
+        assert self.context, "EvaluationExecutor not initialized"
+
+        df_triggers = self.context.tracking_dfs.get(PipelineStage.HANDLE_SINGLE_TRIGGER.name)
+        df_store_models = self.context.tracking_dfs.get(PipelineStage.STORE_TRAINED_MODEL.name)
+        if df_triggers is None or df_store_models is None:
+            logger.warning("Could not run evaluations as no training was found in the tracking info dataframes")
+            return SupervisorLogs(stage_runs=[])
+        df_trainings = df_triggers[["trigger_id", "first_timestamp", "last_timestamp"]].merge(
+            df_store_models[["id_model", "training_id", "trigger_id"]], on="trigger_id"
+        )
+
+        eval_requests: list[EvalRequest] = []
+        for eval_handler in self.eval_handlers:
+            if eval_handler.config.execution_time != "after_pipeline":
+                continue
+
+            handler_eval_requests = eval_handler.get_eval_requests_after_pipeline(df_trainings=df_trainings)
+            eval_requests += handler_eval_requests
+
+        num_workers = None
+        if self.config.supervisor:
+            num_workers = self.config.supervisor.after_pipeline_evaluation_workers
+
+        logs = self._launch_evaluations_async(
+            eval_requests,
+            parent_log=StageLog(
+                id=PipelineStage.EVALUATE_SINGLE.name,
+                start=datetime.datetime.now(),
+                batch_idx=-1,
+                sample_idx=-1,
+                sample_time=-1,
+                trigger_idx=-1,
```

Comment:
We should add a comment explaining this, it at least looks weird. Why don't we care about all of this post pipeline but care about it post training?